internal/snippets/snippets.go

```diff
@@ -0,0 +1,279 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snippets
+
+//go:generate protoc --go_out=. --go_opt=paths=source_relative metadata/metadata.proto
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/googleapis/gapic-generator-go/internal/license"
+	"github.com/googleapis/gapic-generator-go/internal/pbinfo"
+	"github.com/googleapis/gapic-generator-go/internal/snippets/metadata"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+// VersionPlaceholder is the string value $VERSION, intended to be replaced with
+// the actual module version by a generator post-processing script.
+var VersionPlaceholder = "$VERSION"
+
+// headerLen is the length of the Apache license header including trailing newlines.
+var headerLen = len(strings.Split(license.Apache, "\n"))
+
+var spaceSanitizerRegex = regexp.MustCompile(`:\s*`)
+
+var ctxParam = &param{
+	name:  "ctx",
+	pType: "context.Context",
+}
+
+var optsParam = &param{
+	name:  "opts",
+	pType: "...gax.CallOption",
+}
+
+// SnippetMetadata is a model for capturing snippet details and writing them to
+// a snippet_metadata.*.json file.
+type SnippetMetadata struct {
+	// protoPkg is the proto namespace for the API package.
+	protoPkg string
+	// libPkg is the Go import path for the GAPIC client.
+	libPkg string
+	// protoServices is a map of gapic service short names to service structs.
+	protoServices map[string]*service
+	// apiVersion is the gapic service version. (e.g. "v1", "v1beta1")
+	apiVersion string
+	// pkgName is the short package name after the semi-colon in go-gapic-package.
+	pkgName string
+}
+
+// NewMetadata initializes the model that will collect snippet metadata, from:
+// protoPkg - dot-separated, without final type name element (e.g. "google.cloud.bigquery.migration.v2")
+// libPkg - the Go import path for the GAPIC client, per libraryPackage in gapic_metadata.json (e.g. "cloud.google.com/go/bigquery/migration/apiv2")
+// pkgName - stored in g.opts.pkgName, used as an argument to pbinfo.ReduceServName
+func NewMetadata(protoPkg, libPkg, pkgName string) (*SnippetMetadata, error) {
+	protoParts := strings.Split(protoPkg, ".")
+	apiVersion := protoParts[len(protoParts)-1]
```

Comment:
Nit: Maybe just this:
```suggestion
	lastDot := strings.LastIndex(protoPkg, ".")
	apiVersion := protoPkg[lastDot+1:]
```

Saves an intermediate array/slice.

Reply:
Done, thanks!